### ISSUE TITLE
Ability to launch .devcontainer project from command line

### ISSUE THREAD

**Issue by AlJohri** [microsoft/vscode-remote-release#2133]
This issue was closed last time with no resolution: https://github.com/microsoft/vscode-remote-release/issues/1084

How do I use the `code` CLI to launch into the devcontainer? I have a folder with a `.devcontainer` folder. Running `code .` prompts me to re-open as a devcontainer but I want to do it in one shot in my Makefile.

**Comment by hatchermarie2** [comment]
#2141

**Comment by AlJohri** [comment]
I was able to sort of get it to work using the following invocation:

```bash
code --folder-uri vscode-remote://dev-container+2f55736572732f6a6f687269612f446576656c6f706d656e742f70726f71756573742d646f776e6c6f61646572/app
```

which I found by grepping through `cat ~/Library/Application\ Support/Code/storage.json`.

Unfortunately I can't find how to get the mapping of my folder to this random hash: `2f55736572732f6a6f687269612f446576656c6f706d656e742f70726f71756573742d646f776e6c6f61646572`.

**Comment by bhack** [comment]
I am also interested in this to support CI requests. See https://github.com/tensorflow/addons/pull/1309#issuecomment-599123837 /cc @Chuxel

**Comment by Robert-Obermeier** [comment]
> I was able to sort of get it to work using the following invocation:
> 
> ```shell
> code --folder-uri vscode-remote://dev-container+2f55736572732f6a6f687269612f446576656c6f706d656e742f70726f71756573742d646f776e6c6f61646572/app
> ```
> 
> which I found by grepping through `cat ~/Library/Application\ Support/Code/storage.json`.
> 
> Unfortunately I can't find how to get the mapping of my folder to this random hash: `2f55736572732f6a6f687269612f446576656c6f706d656e742f70726f71756573742d646f776e6c6f61646572`.

actually that is your path represented as hex encoded. The upper string decodes to:
`/Users/johria/Development/proquest-downloader`

not sure though why they encode that part.

**Comment by tschaub** [comment]
I'm finding that the "reopen in container" dialog is getting harder and harder to pin down.  Recently when I open my project, I get enough other dialogs about things to do (use Pylance, install a linter, install a test framework) that the "reopen in container" dialog gets hidden until I dismiss some of the other dialogs.

The movie below shows how the dialog shows up initially and then gets hidden.  It is also a bit unsettling to have another dialog/prompt immediately take its place after clicking the "reopen in container" button (but this is another issue).

![jumpy-dialogs](https://user-images.githubusercontent.com/41094/98302375-7f2b9a80-1f79-11eb-9798-eb3f7118cb66.gif)

**Comment by Robert-Obermeier** [comment]
@tschaub I'd open a new issue for this. Cause it is unrelated to this issue here.
You can always see all notifications when clicking on the bell at the lower right corner of the application.
But I do agree that this is not a perfect user experience.

**Comment by takekazuomi** [comment]
I wrote a simple posh script. It works on the Windows side. It doesn't work in wsl2. I don't know why.

```posh
$p = "\\wsl$\Ubuntu\home\takekazu\projects\hoge".ToCharArray() | %{$h=''}{$h += ('{0:x}' -f [int]$_)}{$h}
code --folder-uri "vscode-remote://dev-container+$p/workspaces/hoge"
```

**Comment by Yehonal** [comment]
Based on @takekazuomi script I've managed to create single-line scripts that work on every environments.

Just replace <hostfolder> and <containerfolder> with your absolute paths
NOTE: you can use --file-uri instead of --folder-uri if you want to open a .code-workspace file instead of a folder

## Powershell
```Powershell
$path = "<hostfolder>"; $p = $path.ToCharArray() | %{$h=''}{$h += ('{0:x}' -f [int]$_)}{$h}; code --folder-uri "vscode-remote://dev-container+$p/<containerfolder>"
```

## Bash
```bash
$path = "<hostfolder>"; $p = $path.ToCharArray() | %{$h=''}{$h += ('{0:x}' -f [int]$_)}{$h}; code --folder-uri "vscode-remote://dev-container+$p/<containerfolder>"
```

## WSL

For WSL you can just use the powershell script above using the **"\\wsl$\"** prefixed path, however, if you really need to open the devcontainer within the WSL shell, then you can use this command:

```
/mnt/c/Windows/System32/cmd.exe /C path='<hostfolder>' && p=$(printf "%s" "$path" | xxd -p) && code --folder-uri "vscode-remote://dev-container+${p//[[:space:]]/}/<containerfolder>"
```

**Comment by ilyasotkov** [comment]
@Yehonal Your Bash code is actually PowerShell, you probably made a mistake while copypasting.

For anyone interested in streamlining this as much as possible, I wrote a Python script that fully automates the process of assembling a `--folder-uri` for any directory containing `.devcontainer.json` or `.devcontainer/devcontainer.json`:

https://gist.github.com/ilyasotkov/a5e0bdfeb3bdd7fe5954714387734bdb

You may put the script to e.g. your user home directory, then run

```sh
code --folder-uri $(python3 ~/make_devcontainer_folder_uri.py)
```
or to avoid typing so many characters, create an executable script in your `$PATH` with these contents

```sh
#!/usr/bin/env bash

set -e

uri="$(python3 ~/make_devcontainer_folder_uri.py)"
code --folder-uri $uri
```

**Comment by Yehonal** [comment]
Thanks! I've just fixed it

**Comment by yuusakuri** [comment]
If the GUI `Open Folder in Container` is used, the `workspaceFolder` written in devcontainer.json will be opened, but the above method must be specified separately.

**Comment by fardolieri** [comment]
Thanks to the work of you guys, I created a js script that opens my container and prints the vscode command (in color, why not). I then added it to my `package.json` script section for easy access 

`"container:shortcut": "node ./.devcontainer/open-container-workspace.js",`

I thought I'd share, if it helps someone.

```js
const { exec } = require('child_process');

const printColor = (...args) => console.log('\x1b[36m%s\x1b[0m', ...args);

const workingDirectory = process.cwd().replace(/\\/g, '/');

const hostWorkspaceFileHex = Buffer.from(
  `${workingDirectory}/my.code-workspace`,
).toString('hex');

const containerWorkspaceFile = '/workspace/myProject/my.code-workspace';

const command = `code --file-uri "vscode-remote://dev-container+${hostWorkspaceFileHex}${containerWorkspaceFile}"`;

printColor(command);
exec(command);
```

**Comment by benjaminwood** [comment]
Building off of @Yehonal's bash one-liner shared above :point_up: 

If you're using `zsh`, you can add this function to your `.zshrc`.

```
function code-remote {
  p=$(printf "%s" "$1" | xxd -p) && code --folder-uri "vscode-remote://dev-container+${p//[[:space:]]/}/$2"
}
```

Usage: `code-remote ~/local/path/to/project /remote/path/to/open`

Example: `code-remote ~/projects/foo/bar /app`

Where `~/projects/foo/bar` is the path to the project containing a `.devcontainer` folder and `/app` is the directory I want to open in the container.

Be sure to note what @Yehonal said about .devcontainer folder vs file:

> NOTE: you can use --file-uri instead of --folder-uri if you want to open a .code-workspace file instead of a folder

**Comment by JPustkuchen** [comment]
Hi all and thank you very much for the discussion. I tried several of the examples but couln't make it work. Perhaps someone can help.

I'd like to start vscode from a shell script attached to a running docker container NOT created by a .devcontainer definition. The container is selectable within the VSCode UI, but I can't make it work in bash.
Can I do that by container name? Is there any official documentation about these shell parameters?

For example:
`code --remote dev-container+my_docker_containername_1`

Sorry and thank you for your help!

**Comment by stuartleeks** [comment]
Hi all, I'm happy to report that there is a new CLI released as part of the 0.188.0 release (currently insiders only).

With this version, you can use the "Remote-Containers: Install devcontainer CLI" command to install the CLI. After that you can use `devcontainer-insiders open [path]` to launch a folder as a dev container without the prompt and reload experience. If the `path` argument is omitted it uses the current directory.

**Comment by metaskills** [comment]
> I think this issue can be closed

So when I install the latest NPM package there is no `open` command. Also, I still can not see in that page you listed how to install the insider version. I am assuming that one has the `open` command?

```
> devcontainer --help
devcontainer <command>

Commands:
  devcontainer build [path]  Build a dev container image

Options:
  -h, --help               Show help                                                                           [boolean]
      --disable-telemetry  Disable telemetry                                                  [boolean] [default: false]
```

**Comment by stuartleeks** [comment]
Also, a `run`/`exec` command is being tracked here: https://github.com/microsoft/vscode-remote-release/issues/5416

**Comment by Chuxel** [comment]
@metaskills Unfortunately Codespaces does not support attaching to separate development containers for the same codespace today, so this would not work. You can use Docker Compose to setup side car containers for things like a database, but not attach multiple windows to different containers in the same codespace.  What you can do is have a dev container that covers the requirements for (A) (B) and (C) and then use `.code-workspace` file to get them all to appear in the same window.  I do this in a Codespace I use every day (created from [this repo](https://github.com/chuxel/vscode-dev-containers-scratchpad)).

That said, generally I'd recommend pre-building your images in CI as well just to give you a perf boost and avoid the possibility of something you're depending on being down (e.g. an external packaging system).  Even with small ones its worth doing and the dev container CLI makes it pretty easy to do.

You could in concept assemble a combined image from the dev container, publish it, and then rebuild the Codespace. But, it might actually be something you could do in Actions.  Have a config file that identified the repositories, and an actions workflow that looks for changes to it, then trigger a build and push to the same repository's container registry. The dev container you then actually open just references a finished image.  Just spit-balling an alternate idea.

**Comment by davidpcaldwell** [comment]
Doesn't this already exist? https://code.visualstudio.com/docs/remote/devcontainer-cli

**Comment by BrumGB** [comment]
> Doesn't this already exist? https://code.visualstudio.com/docs/remote/devcontainer-cli

Great, thank you!

**TLD**
1 - Install devcontainer CLI by running vs command palette - Remote-Containers: Install devcontainer CLI
2 - Run command `devcontainer open` instead of `code .` from directory (Or alternatively `devcontainer open *path*`

**Comment by onattech** [comment]
If you are using Docker desktop on Linux, path needs to be like below.
## Bash
 
```shell
path='{"hostPath":"<hostfolder>","localDocker":false,"settings":{"context":"desktop-linux"}}' && p=$(printf "%s" "$path" | xxd -p) && code --folder-uri "vscode-remote://dev-container+${p//[[:space:]]/}<containerfolder>"
 ```

**Comment by jamesdbrock** [comment]
When I have a vscode devcontainer in a repository, I want the README instructions for how to use VS Code in the repository to go like this:

1. Clone the repository.
2. `cd` to the repository directory.
3. Run `code --devcontainer .`.

If there’s a bunch of stuff about first “installing the devcontainer CLI” or “clicking on the Reopen in container pop-up window” then that ruins the simplicity of the instructions.

`code` should have that `--devcontainer` command-line argument.

**Comment by CodeMan99** [comment]
@brad-jones Thank you! I've been wanting use `@devcontainers/cli` instead of the one from `$env:APPDATA\Code\User\globalStorage\ms-vscode-remote.remote-containers\cli-bin\`.

I also made a bash function from your command with a refactor to use `hexdump`. :smiley:

```bash
function devopen() {
    local workspace_folder=$(readlink -f "$1")
    local path_id=$(wslpath -w "$workspace_folder" | hexdump --format '8/1 "%02x"')

    devcontainer up --workspace-folder "$workspace_folder"
    code --folder-uri "vscode-remote://dev-container+${path_id}/workspaces/$(basename $workspace_folder)"
}
```

**Comment by felipecrs** [comment]
Just to mention, the function above only works if the workspace folder is mounted at `/workspaces/${workspaceFolderBasename}`.

**Comment by CodeMan99** [comment]
@felipecrs The value of `workspaceFolder` and `workspaceMount` can be read from the output of `devcontainer read-configuration --workspace-folder .`

```shell
$ devcontainer read-configuration --workspace-folder . | json
[4 ms] @devcontainers/cli 0.30.0. Node.js v18.12.1. linux 5.15.79.1-microsoft-standard-WSL2 x64.
{
  "configuration": {
    "name": "Tunnel Manager",
    "image": "mcr.microsoft.com/devcontainers/javascript-node:0-18-bullseye",
    "customizations": {
      "vscode": {
        "extensions": [
          "EditorConfig.EditorConfig",
          "mads-hartmann.bash-ide-vscode",
          "samverschueren.linter-xo"
        ],
        "settings": {
          "editor.formatOnSave": true,
          "xo.enable": true,
          "xo.format.enable": true,
          "[javascript]": {
            "editor.defaultFormatter": "samverschueren.linter-xo"
          },
          "[typescript]": {
            "editor.defaultFormatter": "samverschueren.linter-xo"
          }
        }
      }
    },
    "remoteUser": "node",
    "features": {
      "ghcr.io/CodeMan99/features/circleci-cli:1": {},
      "ghcr.io/devcontainers/features/docker-in-docker:2": {},
      "ghcr.io/devcontainers/features/git:1": {},
      "ghcr.io/devcontainers/features/github-cli:1": {}
    },
    "configFilePath": {
      "$mid": 1,
      "path": "/home/user/dev/tunnel-manager/.devcontainer/devcontainer.json",
      "scheme": "vscode-fileHost"
    }
  },
  "workspace": {
    "workspaceFolder": "/workspaces/tunnel-manager",
    "workspaceMount": "type=bind,source=/home/user/dev/tunnel-manager,target=/workspaces/tunnel-manager"
  }
}
```

However, I do not have any devcontainer using something other than the default `workspaceFolder`. So I will not be writing that functionality.

**Comment by felipecrs** [comment]
Any ideas on what would be the URI for opening the devcontainer from within a SSH target?

**Comment by felipecrs** [comment]
> Any ideas on what would be the URI for opening the devcontainer from within a SSH target?

BTW, adding `--remote=ssh-remote+<hostname>` to the command does not help (thanks @CodeMan99 for [the suggestion](https://github.com/felipecrs/dotfiles/commit/221fd17652855a5faa2203e9a830d63d59f1b5b1#commitcomment-100739943) though).

**Comment by felipecrs** [comment]
> > Unfortunately I can't find how to get the mapping of my folder to this random hash: `2f55736572732f6a6f687269612f446576656c6f706d656e742f70726f71756573742d646f776e6c6f61646572`.
> 
> actually that is your path represented as hex encoded. The upper string decodes to: `/Users/johria/Development/proquest-downloader`
> 
> seems that splits the part outside and inside the container.
> 
> Fyi for attaching to existing containers on a remote docker host this snippet might help:
> 
> ```python
> remote_description = json.dumps(
>     {"containerName":"/containername","settings":{"host":"ssh://remote-host"}}
> )
> 
> remote_uri = urlunparse(('vscode-remote', f'attached-container+{remote_description.encode("utf8").hex()}', '/path/inside/container', '', '', ''))
> print(remote_uri)
> ```
> 
> I guess that should be most of the stuff one could need at some point. But since Microsoft did not document any of this. I expect this might be subject to change at any point.

Actually I just realized this. I'll try to reimplement it in Bash.

**Comment by michidk** [comment]
I am currently working on a CLI tool for this, written in Rust. You can try it out here: https://github.com/michidk/vscli
It works similar to the bash function from @CodeMan99 but with more error checking and functionalilty. Also, currently expects the workdir inside the container to be at `/workspaces/{workspace name}`. I am working on always getting the configured workspace right now. I'll also add a small TUI that will list the recently opened projects and allow to quick launch them.

**Comment by jamesgecko** [comment]
> > Doesn't this already exist? https://code.visualstudio.com/docs/remote/devcontainer-cli
> 
> Great, thank you!
> 
> **TLDR**
> 
> 1. Install devcontainer CLI by running VS Code command palette - `Remote-Containers: Install devcontainer CLI`
> 2. After installation from run command `devcontainer open` instead of `code .` from directory (Or alternatively `devcontainer open *path*`) to open the current folder in its corresponding dev container.

This has actually been pretty confusing for my team! Some of us installed the `devcontainers` CLI from the VSCode `Remote-Containers` extension. Some of us installed `@devcontainers/cli` via `npm`. They appear to be two versions of the same command with the same version number and different behavior. The `npm` one doesn't have `open`, and requires specifying `--workspace-folder` with `build`.

**Comment by AceHack** [comment]
When can I have `devcontainer open`?  It will be very helpful.

**Comment by anphung** [comment]
@AceHack `devcontainer open` is already available if you install devcontainer via VS Code UI command palette "Remote-Containers: Install devcontainer CLI"

Issue I have with `--folder-uri` is it keeps open "new" session. After open the same folder a few times, in vscode **Remote explorer**, it will shows many of that folder there.

I prefer to use `--remote`. Some change on top of [benjaminwood](https://github.com/benjaminwood)'s [code-remote](https://github.com/microsoft/vscode-remote-release/issues/2133#issuecomment-860863636) to use `--remote`:

```bash
function code_remote {
  p=$(printf "%s" "$1" | xxd -p) && code --remote "dev-container+${p//[[:space:]]/}" "$2"
}
```

Usage: `code-remote ~/local/path/to/project /remote/path/to/open`

Example: `code-remote ~/projects/foo/bar /app`

Where `~/projects/foo/bar` is the path to the project containing a _.devcontainer_ folder and `/app` is the directory you want to open in the container.

This can open both `file` and `folder`.

**Comment by felipecrs** [comment]
For some reason `--remote` doesn't work on Windows when launching from WSL. $2 somehow opens in the local VS Code window rather than opening with the devcontainer window.

**Comment by ecourreges-orange** [comment]
Thanks @brad-jones !  
FWIW, here's a command that works from wsl2 without having to install yet another program like `xxd`, where `hexdump` is mostly preinstalled, also not using variable name "path" which potentially breaks you path (at least on zsh)

```shell
mypath="$(wslpath -w $PWD)" && p=$(printf "%s" "$mypath" | hexdump -v -e '/1 "%02x"') && code --folder-uri "vscode-remote://dev-container+${p}/workspaces/$(basename $PWD)"
```

or as an alias with proper escaping:
```shell
alias devcode='mypath="$(wslpath -w $PWD)" && p=$(printf "%s" "$mypath" | hexdump -v -e "/1 \"%02x\"") && code --folder-uri "vscode-remote://dev-container+${p}/workspaces/$(basename $PWD)"'
```

**Comment by felipecrs** [comment]
I was able to get it working for Remote SSH and Remote Tunnel, the secret is to add `@ssh-remote+hostname` or `@tunnel+tunnelname` after `path_id`.

Finding out such information is hard though. I am parsing the output of `code --status | grep "^Remote:"` to find it.

I put this all together in my code shim:

_Originally posted by @felipecrs in https://github.com/microsoft/vscode-remote-release/issues/5957#issuecomment-2224117990_

I built my own `code` wrapper that has a `--devcontainer` flag. Usage:

```console
code --devcontainer ~/repos/myproject
```

It works for when VS Code is running on Linux, when running from within a remote SSH session (in or out of the vscode integrated terminal) and even within a remote tunnel session.

https://github.com/felipecrs/dotfiles/blob/master/home/dot_local/bin/executable_code
https://github.com/felipecrs/dotfiles/pull/118

It is immensely useful and timesaving for me, so I expect other people to benefit from it too. Simply install it to your `~/.local/bin/code` and you should be good to go.


https://github.com/user-attachments/assets/225db564-3931-46be-8769-c4e70dc274cb

**Comment by a-stankevich** [comment]
Please consider upvoting related VS code issue for `—devcontainer` cli argument implementation. 

https://github.com/microsoft/vscode/issues/236565

**Comment by dacrystal** [comment]
I know this is an older request, but the ability to launch a folder directly into a devcontainer from the command line is exactly what many workflows still lack.

I ran into this same gap and built a small CLI that fills this exact need: you define a devcontainer templates locally and open any project in it from the shell without modifying the repo.

This lets you:

- define devcontainers once
- open any project directly into it (`dev-code open`)
- avoid modifying the repo or copying `.devcontainer/`

So instead of wiring things manually, you can just: `devcode open <template-name> <project>`

Check it out:  https://github.com/dacrystal/dev-code .  I would appreciate a few stars if this is helpful!
